[MS] Adding support for Image Resources and Custom Images with Managed Disks

echuvyrov opened the pull request:
Support for Image ARM Resource and Custom Images with Managed Disks (porting from v0.9.8)

tombuildsstuff commented on website/docs/r/virtual_machine.html.markdown:
```diff
@@ -335,9 +335,11 @@ For more information on the different example configurations, please check out t
-* `publisher` - (Required) Specifies the publisher of the image used to create the virtual machine. Changing this forces a new resource to be created.
-* `offer` - (Required) Specifies the offer of the image used to create the virtual machine. Changing this forces a new resource to be created.
-* `sku` - (Required) Specifies the SKU of the image used to create the virtual machine. Changing this forces a new resource to be created.
+* `image_id` - (Required, when using image resource) Specifies the ID of the (custom) image to use to create the virtual 
```
We'd normally format this as follows:

- `foo` (Optional) Specifies the ID.... Cannot be set when `X`, `Y` and `Z` are set.

And then add an example to show usage of the image_id field?

echuvyrov replied:
Ok, tried to match your style there

tombuildsstuff commented on website/docs/r/image.html.markdown:
```diff
@@ -0,0 +1,92 @@
+* `os_disk_blob_uri` - (Optional) Specifies the URI in Azure storage of the blob that you want to use to create the image.
+* `os_disk_caching` - (Optional) Specifies the caching mode as 'readonly', 'readwrite', or 'none'. The default is none.
+* `os_disk_size_gb` - (Optional) Specifies the size of the image to be created. The target size can't be smaller than the source size.
+* `data_disk` - (Optional) The properties of the data images that 
```
We'd tend to phrase this "One or more `data_disk` elements as defined below"

echuvyrov replied:
done

tombuildsstuff commented on website/docs/r/image.html.markdown:
```diff
@@ -0,0 +1,92 @@
+    the image. Changing this forces a new resource to be created.
+* `location` - (Required) Specified the supported Azure location where the resource exists.
+    Changing this forces a new resource to be created.
+* `source_virtual_machine_id` - (Required when creating image from existing VM) 
```
We'd tend to phrase this as:
```
* `source_virtual_machine_id` - (Optional) The Virtual Machine ID from which to create the image
```

echuvyrov replied:
Tried to match your style there as well

tombuildsstuff commented on website/docs/r/image.html.markdown:
```diff
@@ -0,0 +1,92 @@
+* `source_virtual_machine_id` - (Required when creating image from existing VM) 
+    ID of an existing VM from which the image
+    will be created. VM must be generalized prior to creating image.
+* `os_disk_os_type` - (Required) Specifies the type of operating system contained in the the virtual machine image. Possible values are: Windows or Linux.
```
as above: can we split this out into an `os_disk` block?

echuvyrov replied:
done

tombuildsstuff commented on azurerm/import_arm_image_test.go:
```diff
@@ -0,0 +1,49 @@
+	password := "Password1234!"
+	hostName := "tftestcustomimagesrc"
+	dnsName := fmt.Sprintf("%[1]s.westcentralus.cloudapp.azure.com", hostName)
+	sshPort := "22"
```
I think we can safely default this field within the `testGeneralizeVMImage` function?

echuvyrov replied:
Not sure if we could. I missed the random seed needed to avoid collisions between host names initially, but now that I put it in, the random seed is needed to generalize the VM since the DNS Name is one of the params to that function now. Open to different ideas here as well.

tombuildsstuff replied:
That's true, however we should be able to pass in only the username, password, hostname and region - as we can infer the rest from that?

tombuildsstuff commented on azurerm/resource_arm_image_test.go:
```diff
@@ -0,0 +1,954 @@
+			ssh.Password(password),
+		},
+		//line below for the newer versions of SSH Client
+		//HostKeyCallback: ssh.InsecureIgnoreHostKey(),
```
should we remove this line for the moment?

echuvyrov replied:
done

echuvyrov commented on azurerm/resource_arm_image_test.go:
```diff
@@ -0,0 +1,954 @@
+	return nil
+}
+
+var testAccAzureRMImage_standaloneImage_setup = `
```
done

echuvyrov commented on azurerm/resource_arm_image_test.go:
```diff
@@ -0,0 +1,954 @@
+var testAccAzureRMImage_standaloneImage_setup = `
+resource "azurerm_resource_group" "test" {
+    name = "acctestRG-%[1]d"
+    location = "West Central US"
```
done

echuvyrov commented on azurerm/resource_arm_image_test.go:
done

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,377 @@
+				ConflictsWith: []string{"os_disk_os_type"},
+			},
+
+			"os_disk_os_type": {
```
done

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,377 @@
+
+			"os_disk_managed_disk_id": {
+				Type:     schema.TypeString,
+				Optional: true,
```
done

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,377 @@
+
+			"os_disk_size_gb": {
+				Type:     schema.TypeInt,
+				Optional: true,
```
done

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,377 @@
+		osDisk.Caching = caching
+	}
+
+	diskSize := int32(0)
```
done

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,377 @@
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("[ERROR] Error making Read request on AzureRM Image %s (resource group %s): %s", name, resGroup, err)
```
done

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,377 @@
+	if resp.SourceVirtualMachine != nil {
+		flattenAzureRmSourceVMProperties(d, resp.SourceVirtualMachine)
+	} else if resp.StorageProfile != nil {
+		if resp.StorageProfile.OsDisk != nil {
```
moved

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,377 @@
+	return nil
+}
+
+func flattenAzureRmSourceVMProperties(d *schema.ResourceData, properties *compute.SubResource) {
```
no, removed

echuvyrov commented:
@tombuildsstuff I refactored the dnsName out of individual funcs, let me know if these changes look good to you.

YuriyKischenko commented:
Hi @echuvyrov 
Could you please apply the same changes to the virtual_machine_scale_set as many parts seems to be copied from virtual_machine and image uri is also applicable here.

echuvyrov commented:
Hi @YuriyKischenko - that's the plan :)! I wanted to split the work into smaller chunks, but VMSS support is right behind. I don't think it'll be nearly as much work, since the Image resource would be already done.

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,409 @@
+			"source_virtual_machine_id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"os_disk.os_type"},
```
I used an example from virtual_machine as a reference (such as below) - let me know if I should change:

"managed_disk_id": {
							Type:          schema.TypeString,
							Optional:      true,
							ForceNew:      true,
							Computed:      true,
							ConflictsWith: []string{"storage_os_disk.vhd_uri"},
						},

echuvyrov replied:
removed

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,409 @@
+
+						"os_type": {
+							Type:     schema.TypeString,
+							Optional: true,
```
When I made this field required, I got the following error: 

"1 error(s) occurred:\n\n* resource azurerm_image: source_virtual_machine_id: ConflictsWith cannot contain Required attribute (os_disk.os_type)"}"

So I left the field as optional, but added ForceNew

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,409 @@
+
+						"os_state": {
+							Type:     schema.TypeString,
+							Optional: true,
```
same limitation as above - conflict between "Required" and "ConflictsWith"

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,409 @@
+						"caching": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  "None",
```
changed

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,409 @@
+		}
+	}
+
+	//either source VM or storage profile can be specified, but not both
```
done

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,409 @@
+
+	//either source VM or storage profile can be specified, but not both
+	if (compute.SubResource{}) == sourceVM {
+		properties = compute.ImageProperties{
```
changed

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,409 @@
+	d.Set("resource_group_name", resGroup)
+	d.Set("location", resp.Location)
+
+	//either source VM or storage profile can be specified, but not both
```
changed

echuvyrov commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,409 @@
+			osDisk.OsState = osState
+		}
+
+		managedDisk := &compute.SubResource{}
```
changed

echuvyrov commented on azurerm/resource_arm_image_test.go:
```diff
@@ -0,0 +1,967 @@
+				Destroy: false,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureVMExists("azurerm_virtual_machine.testsource", true),
+					testGeneralizeVMImage(fmt.Sprintf("acctestRG-%[1]d", ri), "testsource",
```
changed

echuvyrov commented on azurerm/resource_arm_virtual_machine.go:
```diff
@@ -1383,18 +1403,32 @@ func expandAzureRmVirtualMachineImageReference(d *schema.ResourceData) (*compute
-		Sku:       &sku,
-		Version:   &version,
-	}, nil
+	if imageReference.ID != nil && imageReference.Publisher != nil {
```
done

echuvyrov commented:
Hi @tombuildsstuff, I made all the changes requested, except for state migration recommendation. I think I need to understand this better - in my mind, I don't think we need state migration since we have a custom hash function. This custom hash function should not change the hash value for existing resources, since the fields for publisher, offer, sku, version should still all be present for non-custom images. Would appreciate you clarifying the thoughts behind the need for state migration please.

Here's the validation I performed for state, let me know if any of those are incorrect:
-Downloaded current public release of TF (0.9.9) and created basic VM + managed disk config (no custom image). Ran 'terraform plan' and the plan is showing what I expected to see. Ran 'terraform apply' to provision infra and persist state.
-Compiled terraform 0.10.0-dev with my changes. Ran 'terraform plan' on the same config - no changes detected, as expected.
-Changed sku in the config, 'terraform plan' shows 1 resource to destroy, 1 to create. As expected, since OS type has 'ForceNew' on it.
-Added the new image resource and changed VM config to use that image resource with custom image. 'terraform plan' shows 3 resources to create, 1 to destroy. I am not sure why 3, since I was expecting just 2 (1 image, 1 VM), but I am not sure if this has to do with hashing/state migration.

Thanks a ton as always!

tombuildsstuff commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,417 @@
+
+						"blob_uri": {
+							Type:     schema.TypeString,
+							Optional: true,
```
is this also computed?

echuvyrov replied:
I don't think it should be - same as blob_uri for os_disk above IMHO

tombuildsstuff commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,417 @@
+						"blob_uri": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
```
is this also computed?

echuvyrov replied:
I don't think this should be computed, if I understood the "Computed" property correctly. In the Developer's guide, "any Resource or Data Source property that should be stored in state but that the user should not be able to configure should have its  Computed  property set to  true."

Per API spec [here](https://docs.microsoft.com/en-us/rest/api/manageddisks/images/images-create) blobUri "specifies the URI in Azure storage of the blob that you want to use to create the virtual machine image" which indicates it should be changeable by the user.

tombuildsstuff commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,417 @@
+						"caching": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  "None",
```
is this also computed?

echuvyrov replied:
[I don't think it should be](https://docs.microsoft.com/en-us/rest/api/manageddisks/images/images-create) - "You can specify the caching mode of the disks as readonly, readwrite, or none, by adding hostCaching. The default is none." Unless I misunderstood the Computed property of course :)

tombuildsstuff commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,417 @@
+
+						"size_gb": {
+							Type:     schema.TypeInt,
+							Optional: true,
```
is this also computed?

echuvyrov replied:
[I don't think it should be](https://docs.microsoft.com/en-us/rest/api/manageddisks/images/images-create) - "you can resize an operating system disk or the data disk in the virtual machine image as you create it by adding diskSizeGB." Unless I misunderstood the Computed property of course :)

tombuildsstuff replied:
so by Computed I mean we will store the remote value in the state regardless of whether it's specified or not (but it depends if it's retuned from the API) :)

tombuildsstuff commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,417 @@
+	//either source VM or storage profile can be specified, but not both
+	if (compute.SubResource{}) == sourceVM {
+		//if both sourceVM and storageProfile are empty, return an error
+		if (compute.ImageStorageProfile{}) == storageProfile {
```
could we simplify this to `if storageProfile.osDisk == nil && len(storageProfile.dataDisks) == 0`?

echuvyrov replied:
changed

tombuildsstuff commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,417 @@
+	}
+
+	//either source VM or storage profile can be specified, but not both
+	if (compute.SubResource{}) == sourceVM {
```
could we simplify this to `if sourceVM.ID == nil`?

echuvyrov replied:
changed

tombuildsstuff commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,417 @@
+	d.Set("location", resp.Location)
+
+	if resp.SourceVirtualMachine != nil {
+		if resp.SourceVirtualMachine.ID != nil {
```
I think we can remove this particular check `*.ID` as it should be handled by d.Set?

echuvyrov replied:
changed

tombuildsstuff commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,417 @@
+		managedDiskID := config["managed_disk_id"].(string)
+		if managedDiskID != "" {
+			managedDisk := &compute.SubResource{}
+			managedDisk.ID = &managedDiskID
```
could we combine these two lines?

echuvyrov replied:
changed

tombuildsstuff commented on azurerm/resource_arm_image.go:
```diff
@@ -0,0 +1,417 @@
+
+		if managedDiskID != "" {
+			managedDisk := &compute.SubResource{}
+			managedDisk.ID = &managedDiskID
```
could we combine these two lines?

echuvyrov replied:
changed

tombuildsstuff commented on azurerm/resource_arm_virtual_machine.go:
```diff
@@ -96,21 +96,27 @@ func resourceArmVirtualMachine() *schema.Resource {
 				MaxItems: 1,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
+						"image_id": {
```
🤔 whilst I'm not opposed to this - would it make more sense to call this `id` to match the API Documentation / ARM Template?

echuvyrov replied:
changed

tombuildsstuff commented on azurerm/resource_arm_virtual_machine.go:
```diff
@@ -1383,18 +1403,32 @@ func expandAzureRmVirtualMachineImageReference(d *schema.ResourceData) (*compute
+		imageID := storageImageRef["image_id"].(string)
+		imageReference = compute.ImageReference{
+			ID: &imageID,
+		}
```
we could simplify this to use the existing imageReference object defined above:

```
if v, ok := storageImageRef["image_id"]; ok {
  imageReference.ID = azure.String(v.(string))
} else {
 ..
}
```

echuvyrov replied:
changed

tombuildsstuff commented on website/azurerm.erb:
```diff
@@ -174,6 +174,15 @@
             </li>
 
+            <li<%= sidebar_current("docs-azurerm-resource-image") %>>
+              <a href="#">Image Resources</a>
```
is it worth placing this in the storage sub-section?

_related:_ I think we should probably reorganise the sidebar in general

echuvyrov replied:
I put it under "Managed Disk" to match the sidebar [here](https://docs.microsoft.com/en-us/rest/api/manageddisks/), hope that's Ok; agree that we should do a bit of re-organization of the side bar

tombuildsstuff commented on website/docs/r/image.html.markdown:
```diff
@@ -0,0 +1,92 @@
+    Changing this forces a new resource to be created.
+* `source_virtual_machine_id` - (Optional) The Virtual Machine ID from which to create the image.
+* `os_disk` - (Optional) One or more os_disk elements as defined below.
+* `data_disk` - (Optional) One or more data_disk elements as defined below.
```
we'd normally wrap this in quote's: ```One or more `data_disk` elements as defined below.```

echuvyrov replied:
changed

echuvyrov commented:
Hey @tombuildsstuff I made the blob_uri, caching and size_gb properties computed and made the tests pass again :). Let me know if the changes look good to you - thanks!

tombuildsstuff commented:
Tests pass (ignoring a subscription-specific test failure): 
<img width="424" alt="screen shot 2017-07-19 at 13 33 56" src="https://user-images.githubusercontent.com/666005/28369676-384fcdc4-6c90-11e7-9809-74a39120d740.png">
<img width="468" alt="screen shot 2017-07-19 at 13 34 00" src="https://user-images.githubusercontent.com/666005/28369680-3b29bcee-6c90-11e7-89a7-198ce8a615a6.png">
